exec/java-exec/src/main/java/org/apache/drill/exec/store/TimedCallable.java

```diff
@@ -0,0 +1,266 @@
+    }
+
+    @Override
+    public V apply(Future<V> future) {
+      Preconditions.checkState(future.isDone());
+      if (!future.isCancelled()) {
+        try {
+          count++;
+          return future.get();
+        } catch (InterruptedException e) {
+          // there is no wait as we are getting result from the completed/done future
+          logger.error("Unexpected exception", e);
+          throw UserException.internalError(e)
+              .message("Unexpected exception")
+              .build(logger);
+        } catch (ExecutionException e) {
+          setThrowable(e.getCause());
+        }
+      } else {
+        setThrowable(new CancellationException());
+      }
+      return null;
+    }
+  }
+
+  private static class Statistics<V> implements Consumer<TimedCallable<V>> {
+    final long start = System.nanoTime();
+    final Stopwatch watch = Stopwatch.createStarted();
+    long totalExecution;
+    long maxExecution;
+    int count;
+    int startedCount;
+    private int doneCount;
+    // measure thread creation times
+    long earliestStart;
+    long latestStart;
+    long totalStart;
+
+    @Override
+    public void accept(TimedCallable<V> task) {
+      count++;
+      long threadStart = task.getStartTime(TimeUnit.NANOSECONDS) - start;
+      if (threadStart >= 0) {
+        startedCount++;
+        earliestStart = Math.min(earliestStart, threadStart);
+        latestStart = Math.max(latestStart, threadStart);
+        totalStart += threadStart;
+        long executionTime = task.getExecutionTime(TimeUnit.NANOSECONDS);
+        if (executionTime != -1) {
+          doneCount++;
+          totalExecution += executionTime;
+          maxExecution = Math.max(maxExecution, executionTime);
+        } else {
+          logger.info("Task {} started at {} did not finish", task, threadStart);
+        }
+      } else {
+        logger.info("Task {} never commenced execution", task);
+      }
+    }
+
+    Statistics<V> collect(final List<TimedCallable<V>> tasks) {
+      totalExecution = maxExecution = 0;
+      count = startedCount = doneCount = 0;
+      earliestStart = Long.MAX_VALUE;
+      latestStart = totalStart = 0;
+      tasks.forEach(this);
+      return this;
+    }
+
+    void info(final String activity, final Logger logger, int parallelism) {
+      if (startedCount > 0) {
+        logger.info("{}: started {} out of {} using {} threads. (start time: min {} ms, avg {} ms, max {} ms).",
+            activity, startedCount, count, parallelism,
+            TimeUnit.NANOSECONDS.toMillis(earliestStart),
+            TimeUnit.NANOSECONDS.toMillis(totalStart) / startedCount,
+            TimeUnit.NANOSECONDS.toMillis(latestStart));
+      } else {
+        logger.info("{}: started {} out of {} using {} threads.", activity, startedCount, count, parallelism);
+      }
+
+      if (doneCount > 0) {
+        logger.info("{}: completed {} out of {} using {} threads (execution time: total {} ms, avg {} ms, max {} ms).",
```

Comment:
What about this info?

Reply:
This is original level from TimedRunnable. If currently there is an excessive logging (I doubt, production system should use WARN or ERROR), please file a JIRA to change it.

Reply:
Sure, but in the future, I'd recommend
- avoid mixing unrelated changes in a single JIRA/PR
- review usage of INFO logging in a separate JIRA/PR
- review production log level also in a separate JIRA/PR (it is uncommon to have INFO level in production currently used by Drill). INFO level is usually used in dev deployments.

Please also see my comment in #1214.